### ISSUE TITLE
[ nc-datadir ] possible bug with USBdevice, nc-automat, nc-format-USB

### ISSUE THREAD

**Issue by Multisaft** [nextcloud/nextcloudpi#1000]
Regarding this topic from nextcloud forum:
https://help.nextcloud.com/t/ncp-cannot-move-datadir/59845/6

is it possible that there is an Issue with the nc-datadir script? does it point not to the device but to the link (myCloudDrive vs. USBdrive).

pasted from the original thread:
OK,
i started from scratch, with 
1. buster lite,
2. apt-get update/upgrade
3. password change of pi user
4. sudo bash
5. raspberrypi installation script ```
curl -sSL https://raw.githubusercontent.com/nextcloud/nextcloudpi/master/install.sh | bash
6. after setup, wizard start for change of data directory to usb storage:
    [ nc-automount ]
    Created symlink /etc/systemd/system/multi-user.target.wants/nc-automount.service → /usr/lib/systemd/system/nc-automount.service.
    Created symlink /etc/systemd/system/multi-user.target.wants/nc-automount-links.service → /usr/lib/systemd/system/nc-automount-links.service.
    automount enabled

    [ nc-format-USB ]
    /dev/sda: 8 bytes were erased at offset 0x00000200 (gpt): 45 46 49 20 50 41 52 54
    /dev/sda: 8 bytes were erased at offset 0xdf90355e00 (gpt): 45 46 49 20 50 41 52 54
    /dev/sda: 2 bytes were erased at offset 0x000001fe (PMBR): 55 aa
    Drive sda formatted successfuly and labeled myCloudDrive

    [ nc-datadir ]
    Maintenance mode enabled
    moving data directory from /var/www/nextcloud/data to /media/USBdrive/ncdata...
    BTRFS filesystem detected
    ERROR: target path already exists: /media/myCloudDrive/ncdata

ls -l /media gives me
total 16

drwxr-xr-x 1 root root 0 Sep 13 12:29 myCloudDrive
lrwxrwxrwx 1 root root 19 Sep 13 12:29 USBdrive -> /media/myCloudDrive

so Link is /media/USBdrive
to Device /media/myCloudDrive
as it was formated to btrfs from [ nc-format-USB ]

the chmod looks good right? read write access for root

if found that discussion on githup from 2017 (quite old I know):
https://github.com/nextcloud/nextcloudpi/issues/305

there someone has the same strange behaviour
I've tried almost everything from nachoparker with removed media symlinks,
changed devicenames
and especially wayneouttheres description of his fix: https://github.com/nextcloud/nextcloudpi/issues/305#issuecomment-377697798

but when I switch to ncp control panel to use nc-datadir with the Data directory:
/media/myCloudDrive/ncdata

it gives:
[ nc-datadir ]
Nextcloud is in maintenance mode - no apps have been loaded

INFO: /media/myCloudDrive/ncdata is not empty. Creating backup /media/myCloudDrive/ncdata-09-13-19
Maintenance mode already enabled
moving data directory from /var/www/nextcloud/data to /media/myCloudDrive/ncdata...
BTRFS filesystem detected
ERROR: target path already exists: /media/myCloudDrive/ncdata

7. the actual status of ncp-report gives:

> <--! Paste this in GitHub report -->

<details>
<summary>NextCloudPi diagnostics</summary>

```

cat: /usr/local/etc/ncp-version: No such file or directory
Nextcloud is in maintenance mode - no apps have been loaded

NextCloudPi version
distribution         Raspbian GNU/Linux 10 \n \l
automount            yes
USB devices          sda 
datadir              /var/www/nextcloud/data
data in SD           yes
data filesystem      ext2/ext3
data disk usage      2.1G/29G
rootfs usage         2.1G/29G
swapfile             /var/swap
dbdir                /var/lib/mysql
Nextcloud check      ok
Nextcloud version    16.0.4.1
HTTPD service        up
PHP service          up
MariaDB service      up
Redis service        up
Postfix service      up
internet check       ok
port check 80        closed
port check 443       closed
IP                   ***REMOVED SENSITIVE VALUE***
gateway              ***REMOVED SENSITIVE VALUE***
interface            eth0
certificates         ***REMOVED SENSITIVE VALUE***
NAT loopback         no
uptime               31min

```
</details>

<details>
<summary>Nextcloud configuration</summary>

```

Nextcloud is in maintenance mode - no apps have been loaded

{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": {
            "0": "localhost",
            "11": "<!DOCTYPE HTML PUBLIC \"-\/\/IETF\/\/DTD HTML 2.0\/\/EN\">\n<html><head>\n<title>302 Found<\/title>\n<\/head><body>\n<h1>Found<\/h1>\n<p>The document has moved <a href=\"https:\/\/icanhazip.com\/\">here<\/a>.<\/p>\n<\/body><\/html>",
            "1": "192.168.1.162",
            "5": "nextcloudpi.local",
            "7": "nextcloudpi",
            "8": "nextcloudpi.lan"
        },
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "16.0.4.1",
        "overwrite.cli.url": "http:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "tempdirectory": "\/var\/www\/nextcloud\/data\/tmp",
        "mail_smtpmode": "sendmail",
        "mail_smtpauthtype": "LOGIN",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "preview_max_x": "2048",
        "preview_max_y": "2048",
        "jpeg_quality": "60",
        "overwriteprotocol": "https",
        "maintenance": true
    }
}

```
</details>

<details>
<summary>HTTPd logs</summary>

```

[Fri Sep 13 11:55:22.480086 2019] [ssl:error] [pid 18866:tid 3069219344] AH02604: Unable to configure certificate localhost:443:0 for stapling
[Fri Sep 13 11:55:22.481142 2019] [mpm_event:notice] [pid 18866:tid 3069219344] AH00489: Apache/2.4.38 (Raspbian) OpenSSL/1.1.1c configured -- resuming normal operations
[Fri Sep 13 11:55:22.481178 2019] [core:notice] [pid 18866:tid 3069219344] AH00094: Command line: '/usr/sbin/apache2'
[Fri Sep 13 12:27:24.283574 2019] [mpm_event:notice] [pid 18866:tid 3069219344] AH00491: caught SIGTERM, shutting down
[Fri Sep 13 12:27:45.314416 2019] [ssl:warn] [pid 631:tid 3069792784] AH01909: localhost:443:0 server certificate does NOT include an ID which matches the server name
[Fri Sep 13 12:27:45.317799 2019] [ssl:error] [pid 631:tid 3069792784] AH02217: ssl_stapling_init_cert: can't retrieve issuer certificate! [subject: CN=raspberrypi / issuer: CN=raspberrypi / serial: 771216E2AA462623F6FC197FB6FED6688F741040 / notbefore: Sep 13 10:50:31 2019 GMT / notafter: Sep 10 10:50:31 2029 GMT]
[Fri Sep 13 12:27:45.317833 2019] [ssl:error] [pid 631:tid 3069792784] AH02604: Unable to configure certificate localhost:443:0 for stapling
[Fri Sep 13 12:27:45.347486 2019] [ssl:warn] [pid 805:tid 3069792784] AH01909: localhost:443:0 server certificate does NOT include an ID which matches the server name
[Fri Sep 13 12:27:45.347805 2019] [ssl:error] [pid 805:tid 3069792784] AH02217: ssl_stapling_init_cert: can't retrieve issuer certificate! [subject: CN=raspberrypi / issuer: CN=raspberrypi / serial: 771216E2AA462623F6FC197FB6FED6688F741040 / notbefore: Sep 13 10:50:31 2019 GMT / notafter: Sep 10 10:50:31 2029 GMT]
[Fri Sep 13 12:27:45.347847 2019] [ssl:error] [pid 805:tid 3069792784] AH02604: Unable to configure certificate localhost:443:0 for stapling
[Fri Sep 13 12:27:45.353170 2019] [mpm_event:notice] [pid 805:tid 3069792784] AH00489: Apache/2.4.38 (Raspbian) OpenSSL/1.1.1c configured -- resuming normal operations
[Fri Sep 13 12:27:45.353263 2019] [core:notice] [pid 805:tid 3069792784] AH00094: Command line: '/usr/sbin/apache2'
[Fri Sep 13 12:28:48.017685 2019] [mpm_event:notice] [pid 805:tid 3069792784] AH00493: SIGUSR1 received.  Doing graceful restart
[Fri Sep 13 12:28:48.042836 2019] [ssl:warn] [pid 805:tid 3069792784] AH01909: localhost:4443:0 server certificate does NOT include an ID which matches the server name
[Fri Sep 13 12:28:48.043021 2019] [ssl:error] [pid 805:tid 3069792784] AH02217: ssl_stapling_init_cert: can't retrieve issuer certificate! [subject: CN=raspberrypi / issuer: CN=raspberrypi / serial: 771216E2AA462623F6FC197FB6FED6688F741040 / notbefore: Sep 13 10:50:31 2019 GMT / notafter: Sep 10 10:50:31 2029 GMT]
[Fri Sep 13 12:28:48.043039 2019] [ssl:error] [pid 805:tid 3069792784] AH02604: Unable to configure certificate localhost:4443:0 for stapling
[Fri Sep 13 12:28:48.044091 2019] [mpm_event:notice] [pid 805:tid 3069792784] AH00489: Apache/2.4.38 (Raspbian) OpenSSL/1.1.1c configured -- resuming normal operations
[Fri Sep 13 12:28:48.044111 2019] [core:notice] [pid 805:tid 3069792784] AH00094: Command line: '/usr/sbin/apache2'
[Fri Sep 13 12:29:07.205573 2019] [proxy_fcgi:error] [pid 1558:tid 2755716128] [client 192.168.1.181:50067] AH01071: Got error 'PHP message: PHP Notice:  Undefined index: app in /var/www/ncp-web/index.php on line 238PHP message: PHP Notice:  Undefined index: app in /var/www/ncp-web/index.php on line 244'
[Fri Sep 13 12:31:17.467573 2019] [proxy_fcgi:error] [pid 1558:tid 2822923296] [client 192.168.1.181:50067] AH01071: Got error 'PHP message: PHP Notice:  Undefined index: app in /var/www/ncp-web/index.php on line 238PHP message: PHP Notice:  Undefined index: app in /var/www/ncp-web/index.php on line 244'

```
</details>

<details>
<summary>Database logs</summary>

```

2019-09-13 12:27:46 0 [Note] InnoDB: Uses event mutexes
2019-09-13 12:27:46 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2019-09-13 12:27:46 0 [Note] InnoDB: Number of pools: 1
2019-09-13 12:27:46 0 [Note] InnoDB: Using generic crc32 instructions
2019-09-13 12:27:46 0 [Note] InnoDB: Initializing buffer pool, total size = 1.625G, instances = 1, chunk size = 128M
2019-09-13 12:27:46 0 [Note] InnoDB: Completed initialization of buffer pool
2019-09-13 12:27:46 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
2019-09-13 12:27:46 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
2019-09-13 12:27:46 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2019-09-13 12:27:46 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2019-09-13 12:27:46 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
2019-09-13 12:27:46 0 [Note] InnoDB: 10.3.17 started; log sequence number 6801485; transaction id 5149
2019-09-13 12:27:46 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
2019-09-13 12:27:46 0 [Note] Plugin 'FEEDBACK' is disabled.
2019-09-13 12:27:46 0 [Note] Server socket created on IP: '127.0.0.1'.
2019-09-13 12:27:46 0 [Note] Reading of all Master_info entries succeeded
2019-09-13 12:27:46 0 [Note] Added new Master_info '' to hash table
2019-09-13 12:27:46 0 [Note] /usr/sbin/mysqld: ready for connections.
Version: '10.3.17-MariaDB-0+deb10u1'  socket: '/run/mysqld/mysqld.sock'  port: 3306  Raspbian 10
2019-09-13 12:27:47 0 [Note] InnoDB: Buffer pool(s) load completed at 190913 12:27:47

```
</details>

<details>
<summary>Nextcloud logs</summary>

```


```
</details>

**Comment by FuerteLoco** [comment]
I can confirm, that there is something wrong (see [nc-datadir.sh](https://github.com/nextcloud/nextcloudpi/blob/master/bin/ncp/CONFIG/nc-datadir.sh)): `DATADIR` is always created and it is used to determine the "real" path name (instead of the symlink). But if you are using BTRFS, then the subvolume cannot be created, since it is the same as the directory created before:

```bash
  # resolve symlinks and use the real path
  mkdir -p "$DATADIR"
  DATADIR=$(cd "$DATADIR" && pwd -P)

  # use subvolumes, if BTRFS
  [[ "$( stat -fc%T "$BASEDIR" )" == "btrfs" ]] && {
    echo "BTRFS filesystem detected"
    btrfs subvolume create "$DATADIR" || return  1
  }
```
I made a temporary workaround and inserted `rm -r "$DATADIR"` before the subvolume creation. Now it started copying my files. Fingers crossed that nothing goes wrong.

**Comment by JDC8077** [comment]
I am having the same, using last NCP OdroidHC2 image, updated to 1.16.5.

Currently blocked on new install, Linux skills weak so uncertain about how to apply the workaround above.

**Comment by nachoparker** [comment]
I pushed a fix in `v1.16.6`. Please confirm that it works for you

**Comment by JDC8077** [comment]
Many Thanks 

> 
> 
> I pushed a fix in `v1.16.6`. Please confirm that it works for you
